src/libsyntax/ext/expand.rs

```diff
@@ -278,13 +266,48 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
             self.cx.current_expansion = invoc.expansion_data.clone();
 
             self.cx.current_expansion.mark = scope;
-            let expansion = match ext {
-                Some(ext) => self.expand_invoc(invoc, ext),
-                None => invoc.expansion_kind.dummy(invoc.span()),
+            let (expansion, new_invocations) = if let Some(ext) = ext {
```

Comment:
Probably -- I was thinking maybe a separate `collect_derives`. I was thinking once we implement inert attribute vs macro attribute ambiguity errors (i.e. scope (b) from https://github.com/rust-lang/rust/issues/39347#issuecomment-275828264) and expanding items before derives, we'll have a better idea of how best to abstract / refactor out.